File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.js
function solution(arr) {
    var answer = [0, 0];
    
    function compress(x, y, size) {
        let first = arr[x][y];
        
        // 주어진 영역이 하나의 숫자로 이루어져 있는지 확인
        for (let i = x; i < x + size; i++) {
            for (let j = y; j < y + size; j++) {
                // 다른 숫자가 있는 경우 영역 분할
                if (arr[i][j] !== first) { 
                    let newSize = size / 2;
                    compress(x, y, newSize);            // 좌상
                    compress(x, y + newSize, newSize);  // 우상
                    compress(x + newSize, y, newSize);  // 좌하
                    compress(x + newSize, y + newSize, newSize); // 우하
                    return;
                }
            }
        }
        
        // 모든 숫자가 같다면 해당 숫자 카운트 증가
        answer[first]++;
    }
    
    compress(0, 0, arr.length);
    
    return answer;
}

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
def solution(arr):
    
    def compress(x, y, n):
        # 모든 값이 같은지 확인
        flag = True
        base = arr[x][y]
        for i in range(x, x + n):
            for j in range(y, y + n):
                if arr[i][j] != base:
                    flag = False
        # if 같으면: 해당 값의 카운트 리턴
        if flag:
            return [1, 0] if base == 0 else [0, 1] 
        # else: 다르면 4분할 후 재귀 호출
        else:
            half = n // 2
            lu0, lu1 = compress(x, y, half)
            ru0, ru1 = compress(x, y + half , half)
            ld0, ld1 = compress(x + half, y, half)
            rd0, rd1 = compress(x + half, y + half, half)
            
            # 각각의 리턴값을 합침
            return [lu0+ru0+ld0+rd0, lu1+ru1+ld1+rd1]
        
    return compress(0, 0, len(arr))